src/transformers/trainer.py

```diff
@@ -3888,53 +3894,66 @@ def evaluation_loop(
                 logits = self.gather_function((logits))
-                if not self.args.batch_eval_metrics or description == "Prediction":
-                    all_preds.add(logits)
             if labels is not None:
                 labels = self.gather_function((labels))
-                if not self.args.batch_eval_metrics or description == "Prediction":
-                    all_labels.add(labels)
-
-            self.control = self.callback_handler.on_prediction_step(args, self.state, self.control)
-
-            if self.args.batch_eval_metrics:
-                if self.compute_metrics is not None and logits is not None and labels is not None:
-                    is_last_step = self.accelerator.gradient_state.end_of_dataloader
-                    if args.include_inputs_for_metrics:
-                        metrics = self.compute_metrics(
-                            EvalPrediction(predictions=logits, label_ids=labels, inputs=inputs),
-                            compute_result=is_last_step,
-                        )
-                    else:
-                        metrics = self.compute_metrics(
-                            EvalPrediction(predictions=logits, label_ids=labels),
-                            compute_result=is_last_step,
-                        )
 
-                del losses, logits, labels, inputs
-                torch.cuda.empty_cache()
+            # Collect inputs, predictions and labels for the whole epoch
+            all_losses.add(losses)
+            if do_accumulate_epoch_data:
+                all_inputs.add(inputs_decode)
+                all_preds.add(logits)
+                all_labels.add(labels)
+
+            # Put tesnors on the CPU if we have done enough accumulation steps to save GPU memory
+            if args.eval_accumulation_steps is not None and (step + 1) % args.eval_accumulation_steps == 0:
+                all_losses.cpu()
+                all_preds.cpu()
+                all_labels.cpu()
+                all_inputs.cpu()
+
+            # Compute metrics on batch if `training_args.batch_eval_metrics=True`
+            if (
+                args.batch_eval_metrics
+                and self.compute_metrics is not None
+                and logits is not None
+                and labels is not None
+            ):
+                eval_batch_inputs = None
+                eval_batch_logits = None
+                eval_batch_labels = None
+
+                to_numpy = args.eval_numpify_tensors
+
+                if inputs_decode is not None:
+                    eval_batch_inputs = nested_numpify(inputs_decode) if to_numpy else nested_cpu(inputs_decode)
+                if logits is not None:
+                    eval_batch_logits = nested_numpify(logits) if to_numpy else nested_cpu(logits)
+                if labels is not None:
+                    eval_batch_labels = nested_numpify(labels) if to_numpy else nested_cpu(labels)
```

Comment:
We can probably refactor this as:

(Preference refactor I think)

```suggestion
                nested_convert_func = nested_numpify if args.eval_numpify_tensors else nested_cpu

                if inputs_decode is not None:
                    eval_batch_inputs = nested_convert_func (inputs_decode)
                if logits is not None:
                    eval_batch_logits = nested_convert_func (logits)
                if labels is not None:
                    eval_batch_labels = nested_convert_func (labels)
```